### ISSUE TITLE
Realm: incorrect cuda synchronization in non-hijack mode

### ISSUE THREAD

**Issue by bandokihiro** [StanfordLegion/legion#981]
@streichler knows what this is about (and I don't know enough to provide an useful description).

**Comment by streichler** [comment]
The problem is that we do a `cuCtxSynchronize`, which prevents that GPU processor from running any other tasks, but it doesn't stop the previous task from claiming it's complete, which means copies to/from device memory might not be properly synchronized.  This patch works around it:
```
diff --git a/runtime/realm/cuda/cuda_module.cc b/runtime/realm/cuda/cuda_module.cc
index 8c8f555..8d96cef 100644
--- a/runtime/realm/cuda/cuda_module.cc
+++ b/runtime/realm/cuda/cuda_module.cc
@@ -1666,9 +1666,6 @@ namespace Realm {
         ThreadLocal::created_gpu_streams = 0;
       }
 
-      // now enqueue the fence on the local stream
-      fence->enqueue_on_stream(s);
-
       // A useful debugging macro
 #ifdef FORCE_GPU_STREAM_SYNCHRONIZE
       CHECK_CU( cuStreamSynchronize(s->get_stream()) );
@@ -1693,6 +1690,9 @@ namespace Realm {
       // no hijack at all, so always synchronize to ensure all effects are observable
       CHECK_CU( cuCtxSynchronize() );
 #endif
+      // now enqueue the fence on the local stream
+      fence->enqueue_on_stream(s);
+
       // pop the CUDA context for this GPU back off
       gpu_proc->gpu->pop_context();
 ```
but the right thing to do is to farm the `cuCtxSynchronize` calls off to subthreads to make them behave like `cuEventRecord`.

**Comment by streichler** [comment]
This is fixed in master with commit 68698248.
